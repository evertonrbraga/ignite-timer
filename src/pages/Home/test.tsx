import { fireEvent, render, screen, waitFor } from '@testing-library/react'
import { Home } from '.'
import { ThemeProvider } from 'styled-components'
import { defaultTheme } from 'styles/themes/default'

describe('<Home />', () => {
  it('should render the heading', () => {
    const { container } = render(
      <ThemeProvider theme={defaultTheme}>
        <Home />
      </ThemeProvider>
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  beforeEach(() => {
    render(
      <ThemeProvider theme={defaultTheme}>
        <Home />
      </ThemeProvider>
    )
  })

  it('should check if static texts are displayed', () => {
    const label01 = screen.getByText('Vou trabalhar em')
    const label02 = screen.getByText('durante')
    const label03 = screen.getByText('minutos.')

    expect(label01).toBeInTheDocument()
    expect(label02).toBeInTheDocument()
    expect(label03).toBeInTheDocument()
  })

  it("should check if the project name's input is working properly", () => {
    const input = screen.getByPlaceholderText('Dê um nome para o seu projeto')
    const dataList = screen.getByLabelText('lista de sugestões de projeto')

    expect(dataList).toBeInTheDocument()
    expect(input).toHaveAttribute('id', 'task')
    expect(input).toHaveAttribute('list', 'task-suggestions')

    expect(input).toHaveValue('')
    fireEvent.change(input, { target: { value: 'Some text' } })
    expect(input).toHaveValue('Some text')
  })

  it("should check if the time's input is working properly", () => {
    const input = screen.getByPlaceholderText('00')

    expect(input).toHaveAttribute('type', 'number')
    expect(input).toHaveAttribute('step', '5')
    expect(input).toHaveAttribute('min', '5')
    expect(input).toHaveAttribute('max', '60')

    fireEvent.focus(input)
    fireEvent.change(input, { target: { value: '22' } })
    fireEvent.keyDown(input, { key: 'ArrowUp' })
    waitFor(() => expect(input).toHaveValue('25'))
    fireEvent.keyDown(input, { key: 'ArrowDown' })
    waitFor(() => expect(input).toHaveValue('20'))
  })

  it('should check if the countdown has the right styles', () => {
    const countdownContainer = screen.getByLabelText('container do contador')
    const number = screen.getByLabelText('número do contador')
    const separator = screen.getByLabelText('separador')

    expect(countdownContainer).toHaveStyle(`
      font-family: 'Roboto Mono',monospace;
      font-size: 10rem;
      line-height: 8rem;
      color: #E1E1E6;
      display: flex;
      gap: 1rem;
    `)

    expect(number).toHaveStyle(`
      background: #29292E;
      padding: 2rem 1rem;
      border-radius: 8px;
    `)

    expect(separator).toHaveStyle(`
      padding: 2rem 0;
      color: #00875F;
      width: 4rem;
      overflow: hidden;
      display: flex;
      justify-content: center;
    `)
  })

  it('should check if the disable feature works properly', () => {
    const taskInput = screen.getByPlaceholderText(
      'Dê um nome para o seu projeto'
    )
    const minutesAmountInput = screen.getByPlaceholderText('00')
    const button = screen.getByText('Começar')

    expect(button).toBeDisabled()

    fireEvent.change(taskInput, { target: { value: 'First project' } })
    fireEvent.change(minutesAmountInput, { target: { value: 25 } })

    expect(button).not.toBeDisabled()
  })

  it('should check if the form fields are resetted on submit', () => {
    const taskInput = screen.getByPlaceholderText(
      'Dê um nome para o seu projeto'
    )
    const minutesAmountInput = screen.getByPlaceholderText('00')
    const button = screen.getByText('Começar')

    fireEvent.change(taskInput, { target: { value: 'First project' } })
    fireEvent.change(minutesAmountInput, { target: { value: 25 } })

    fireEvent.click(button)

    waitFor(() => expect(taskInput).toHaveValue(''))
    waitFor(() => expect(minutesAmountInput).toHaveValue(0))
  })
})
