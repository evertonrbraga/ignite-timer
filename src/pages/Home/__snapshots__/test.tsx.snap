// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should render the heading 1`] = `
.c7 {
  width: 100%;
  border: 0;
  padding: 1rem;
  border-radius: 8px;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0.5rem;
  font-weight: bold;
  cursor: pointer;
  background: #00875F;
  color: #E1E1E6;
  opacity: 0.7;
  cursor: not-allowed;
}

.c0 {
  flex: 1;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.c0 form {
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 3.5rem;
}

.c1 {
  width: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  gap: 0.5rem;
  color: #E1E1E6;
  font-size: 1.125rem;
  font-weight: bold;
  flex-wrap: wrap;
}

.c2 {
  background: transparent;
  height: 2.5rem;
  border: 0;
  border-bottom: 2px solid #7C7C8A;
  font-weight: bold;
  font-size: 1.125rem;
  padding: 0 0.5rem;
  color: #E1E1E6;
}

.c2:focus {
  box-shadow: none;
  border-color: #00875F;
}

.c2::placeholder {
  color: #7C7C8A;
}

.c3 {
  flex: 1;
}

.c3::-webkit-calendar-picker-indicator {
  display: none!important;
}

.c4 {
  width: 4rem;
}

.c5 {
  font-family: 'Roboto Mono',monospace;
  font-size: 10rem;
  line-height: 8rem;
  color: #E1E1E6;
  display: flex;
  gap: 1rem;
}

.c5 span {
  background: #29292E;
  padding: 2rem 1rem;
  border-radius: 8px;
}

.c6 {
  padding: 2rem 0;
  color: #00875F;
  width: 4rem;
  overflow: hidden;
  display: flex;
  justify-content: center;
}

<main
  class="c0"
>
  <form
    action=""
  >
    <div
      class="c1"
    >
      <label
        for="task"
      >
        Vou trabalhar em
      </label>
      <input
        class="c2 c3"
        id="task"
        list="task-suggestions"
        name="task"
        placeholder="Dê um nome para o seu projeto"
      />
      <datalist
        aria-label="lista de sugestões de projeto"
        id="task-suggestions"
      >
        <option
          value="Projeto 1"
        />
        <option
          value="Projeto 2"
        />
        <option
          value="Projeto 3"
        />
        <option
          value="Banana"
        />
      </datalist>
      <label
        for="minutesAmount"
      >
        durante
      </label>
      <input
        class="c2 c4"
        id="minutesAmount"
        max="60"
        min="5"
        name="minutesAmount"
        placeholder="00"
        step="5"
        type="number"
      />
      <span>
        minutos.
      </span>
    </div>
    <div
      aria-label="container do contador"
      class="c5"
    >
      <span
        aria-label="número do contador"
      >
        0
      </span>
      <span>
        0
      </span>
      <div
        aria-label="separador"
        class="c6"
      >
        :
      </div>
      <span>
        0
      </span>
      <span>
        0
      </span>
    </div>
    <button
      class="c7"
      disabled=""
    >
      <svg
        aria-label="ícone de play"
        fill="currentColor"
        height="24"
        viewBox="0 0 256 256"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="none"
          height="256"
          width="256"
        />
        <path
          d="M228.1,121.2,84.2,33.2A8,8,0,0,0,72,40V216a8,8,0,0,0,12.2,6.8l143.9-88A7.9,7.9,0,0,0,228.1,121.2Z"
          fill="none"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="16"
        />
      </svg>
      Começar
    </button>
  </form>
</main>
`;
